src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/sql.js

```diff
@@ -0,0 +1,274 @@
+                sum = (sum + 31 * s.charCodeAt(i)) % 255
+            }
+            return sum
+        }
+        function getColor(name) {
+            return [hash('r' + name), hash('g' + name), hash('b' + name), 1]
+        }
+        this.theme = {
+            getColorForType: getColor,
+            getForegroundColorForType: function (x) {
+                return [255, 255, 255, 1]
+            },
+        }
+    }
+
+    onDataReceived(data) {
+        while (this.dom.firstChild) {
+            this.dom.removeChild(this.dom.lastChild)
+        }
+        let parsedData = data
+        if (typeof data === 'string') {
+            parsedData = JSON.parse(data)
+        }
+
+        const style = `
+        <style>
+        .sql {
+            font-family: DejaVuSansMonoBook, sans-serif;
+            font-size: 12px;
+            margin-left: 7px;
+        }
+        .interpolation {
+            border-radius: 6px;
+            padding:1px 2px 1px 2px;
+            display: inline;
+        }
+        .mismatch-parent {
+            position: relative;
+            display: inline-flex;
+            justify-content: center;
+        }
+        .mismatch-mouse-area {
+            display: inline;
+            position: absolute;
+            width: 150%;
+            height: 150%;
+            align-self: center;
+            z-index: 0;
+        }
+        .mismatch {
+            z-index: 1;
+        }
+        .modulepath {
+            color: rgba(150, 150, 150, 0.9);
+        }
+        .tooltip {
+            font-family: DejaVuSansMonoBook, sans-serif;
+            font-size: 12px;
+            opacity: 0;
+            transition: opacity 0.2s;
+            display: inline-block;
+            white-space: nowrap;
+            background-color: rgba(249, 249, 249, 1);
+            box-shadow: 0 0 16px rgba(0, 0, 0, 0.16);
+            text-align: left;
+            border-radius: 6px;
+            padding: 5px;
+            position: absolute;
+            z-index: 99999;
+            pointer-events: none;
+        }
+        </style>
+        `
+
+        function renderColor(color) {
+            return 'rgba(' + color[0] + ',' + color[1] + ',' + color[2] + ',' + color[3] + ')'
+        }
+
+        const bgAlpha = 0.25
+        const theme = this.theme
+
+        function splitTypeName(name) {
+            var ix = name.lastIndexOf('.')
+            if (ix < 0) {
+                return ['', name]
+            }
+
+            return [name.substr(0, ix + 1), name.substr(ix + 1)]
+        }
+
+        function renderParameter(param) {
+            const actualType = param.actual_type
+            let value = param.value
+
+            if (actualType == 'Builtins.Main.Text') {
+                value = "'" + value.replaceAll("'", "''") + "'"
+            }
+
+            const actualTypeColor = theme.getColorForType(actualType)
+            const fgColor = actualTypeColor
+            let bgColor = [...fgColor.slice(0, 3), bgAlpha]
```

Comment:
magic numbers. Do not use literals like that. Extract them to well named variables. You can use `0` as it means start, but `3` is magical

Reply:
Ok, I changed this to `const baseChannels = 3` since that `3` represents the `RGB` channels that are kept when changing opacity; it's also now a separate function so it is easier to see what it does.

I assume that I can use `2` without treating it like a magic number when I'm dividing things by two, right? (Like I want to compute the middle of something, so I have `width / 2`)